pubsub/trace.go

```diff
@@ -306,71 +306,118 @@ func (c messageCarrier) Keys() []string {
 	return out
 }
 
+// injectPropagation injects context data into the Pub/Sub message's Attributes field.
+func injectPropagation(ctx context.Context, msg *Message) {
+	// only inject propagation if a valid span context was detected.
+	if trace.SpanFromContext(ctx).SpanContext().IsValid() {
+		if msg.Attributes == nil {
+			msg.Attributes = make(map[string]string)
+		}
+		propagation.TraceContext{}.Inject(ctx, newMessageCarrier(msg))
+	}
+}
+
 const (
-	// span names
-	publisherSpanName          = "send"
-	publishFlowControlSpanName = "publisher flow control"
-	publishBatcherSpanName     = "publish batcher"
-	publishRPCSpanName         = "publish"
+	// publish span names
+	createSpanName     = "create"
+	publishFCSpanName  = "publisher flow control"
+	batcherSpanName    = "publisher batching"
+	publishRPCSpanName = "publish"
+
+	// subscribe span names
+	subscribeSpanName = "subscribe"
+	ccSpanName        = "subscriber concurrency control"
+	processSpanName   = "process"
+	scheduleSpanName  = "subscribe scheduler"
+	modackSpanName    = "modack"
+	ackSpanName       = "ack"
+	nackSpanName      = "nack"
+
+	// event names
+	eventPublishStart = "publish start"
+	eventPublishEnd   = "publish end"
+	eventModackStart  = "modack start"
+	eventModackEnd    = "modack end"
+	eventAckStart     = "ack start"
+	eventAckEnd       = "ack end"
+	eventNackStart    = "nack start"
+	eventNackEnd      = "nack end"
+	eventAckCalled    = "ack called"
+	eventNackCalled   = "nack called"
+
+	resultAcked   = "acked"
+	resultNacked  = "nacked"
+	resultExpired = "expired"
 
 	// custom pubsub specific attributes
-	numBatchedMessagesAttribute = "messaging.pubsub.num_messages_in_batch"
-	orderingAttribute           = "messaging.pubsub.ordering_key"
+	gcpProjectIDAttribute    = "gcp.project_id"
+	pubsubPrefix             = "messaging.gcp_pubsub."
+	orderingAttribute        = pubsubPrefix + "message.ordering_key"
+	deliveryAttemptAttribute = pubsubPrefix + "message.delivery_attempt"
+	eosAttribute             = pubsubPrefix + "exactly_once_delivery"
+	ackIDAttribute           = pubsubPrefix + "message.ack_id"
+	resultAttribute          = pubsubPrefix + "result"
+	receiptModackAttribute   = pubsubPrefix + "is_receipt_modack"
```

Comment:
I like the way you've done this. I should restructure some of this in the Node implementation.